pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingTable.java

```diff
@@ -18,24 +18,46 @@
  */
 package org.apache.pinot.core.routing;
 
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.pinot.core.transport.ServerInstance;
 
 
 public class RoutingTable {
-  private final Map<ServerInstance, List<String>> _serverInstanceToSegmentsMap;
+  // Optional segments are those not online as in ExternalView but might have been online on servers already, e.g.
+  // the newly created consuming segments. Such segments were simply skipped by brokers at query routing time, but that
+  // had caused wrong query results, particularly for upsert tables. Instead, we should pass such segments to servers
+  // and let them decide how to handle them, e.g. skip them upon issues or include them for better query results.
+  private final Map<ServerInstance, Pair<List<String>/*non-optional segments*/, List<String>/*optional segments*/>>
+      _serverInstanceToSegmentsMap;
   private final List<String> _unavailableSegments;
   private final int _numPrunedSegments;
 
   public RoutingTable(Map<ServerInstance, List<String>> serverInstanceToSegmentsMap, List<String> unavailableSegments,
       int numPrunedSegments) {
-    _serverInstanceToSegmentsMap = serverInstanceToSegmentsMap;
+    this(serverInstanceToSegmentsMap, Collections.emptyMap(), unavailableSegments, numPrunedSegments);
+  }
+
+  public RoutingTable(Map<ServerInstance, List<String>> serverInstanceToSegmentsMap,
```

Comment:
Suggest keeping this special handling at the caller side. `RoutingTable` should be a thin wrapper class without any complex logic